crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -402,24 +399,52 @@ def upgrade_drbd_cluster(cluster)
         # First node to upgrade is DRBD slave. There might be more resources using DRBD backend
         # but the Master/Slave distribution might be different among them.
         # Therefore, we decide only by looking at the first DRBD resource we find in the cluster.
-        drbd_slave = ""
-        drbd_master = ""
-
-        drbd_nodes.each do |drbd_node|
-          cmd = "LANG=C crm resource status ms-drbd-{postgresql,rabbitmq}\
-          | sort | head -n 2 | grep \\$(hostname) | grep -q Master"
-          out = drbd_node.run_ssh_cmd(cmd)
-          if out[:exit_code].zero?
-            drbd_master = drbd_node.name
+        first = ""
+        second = ""
+
+        # But we have to make sure that when this method is invoked for a second time
+        # (probably after previous failure), same order of nodes is selected as in the first case.
+        # Looking at the DRBD state could be problematic, because master and slave are normally
+        # switched during the upgrade. So let's look at the data we have about the upgraded nodes.
+        status = ::Crowbar::UpgradeStatus.new
+        # current_node is only set when the upgrade of a node starts, and is empty at the start.
+        current_node = status.progress[:current_node] || {}
```

Comment:
There might be a problem here, as just discussed with @rhafer : this `current_node` might not belong to the DRBD cluster we're trying to upgrade. We should check for the nodes state at the beggining of the method.